Update article.md (#21)
javascript-tutorial/de.javascript.info: 1-js/02-first-steps/14-function-basics/article.md

# Funktionen (*Functions*)

Des öfteren kommt es vor dass wir eine (ähnliche) Aktion an mehreren Stellen im *script* ausführen möchten. 

> Z. B. wenn wir dem Besucher unsere Seite eine schöne Nachreicht anzeigen wollen wenn dieser sich ein-, auslogt oder eventuell noch bei anderen Events. 

Reviewer: ```suggestion
Z. B. wenn wir dem Besucher unsere Seite eine schöne Nachricht anzeigen wollen wenn dieser sich einloggt, ausloggt und eventuell noch bei anderen Events. 
```

> Funktionen kann man auch als das "Hauptgerüst" des Programs verstehen. Sie erlauben es uns geschriebenen Code häufig abzurufen ohne diesen erneut schreiben zu müssen.

Reviewer: ```suggestion
Funktionen kann man auch als das "Bausteine" des Programs verstehen. Sie erlauben es uns geschriebenen Code häufig abzurufen ohne diesen erneut schreiben zu müssen.
```

> Wir wurden bereits Zeugen von eingebauten (*built-in*) Funktionen (*functions*) wie `alert(nachricht)`, `prompt(nachricht, default)` und `confirm(frage)`. Es ist uns aber auch möglich slebst Funktionen zu kreieren.

Reviewer: ```suggestion
Wir wurden bereits Zeugen von eingebauten (*built-in*) Funktionen (*functions*) wie `alert(message)`, `prompt(message, default)` und `confirm(question)`. Es ist uns aber auch möglich selbst Funktionen zu kreieren.
```

## Funktionsdeklarierung (*Function Declaration*)

Funktionen können wir m. H. von Funktionsdeklarierungen (*function declaration*) erstellen.

Das sieht wie folgt aus:

```js
function showMessage() {
  alert( 'Hello everyone!' );
}
```

> Das *keyword* (Schlüsselwort) `function` steht an erster Stelle, darauf folgt der *Name der Funktion*. Dann noch eine Liste an Parameter zwischen den Paranthesen (mit einem "," voneinander getrennt; om obigen Beispiel leer) und letztlich der Code der Funktion, der auch als "Körper der Funktion" (*the function body*) genannt wird, der zwischen den eckigen Klammern steht (*curly brackets*).

Author: ```suggestion
Das *keyword* (Schlüsselwort) `function` steht an erster Stelle, darauf folgt der *Name der Funktion*. Dann noch eine Liste an Parameter zwischen den Parenthesen (mit einem "," voneinander getrennt; im obigen Beispiel leer) und letztlich der Code der Funktion, der auch als "Körper der Funktion" (*the function body*) genannt wird, der zwischen den Akkoladen (*curly braces*) steht.
```

```js
function name(parameters) {
  ...body...
}
```

> Unsere neue Funktion kann mittels dessen Name abgerufen werden (*called*): `showMessage()`.

Reviewer 2: ```suggestion
Unsere neue Funktion kann mit ihrem Namen abgerufen werden: `showMessage()`.
```

> Bspw. :

Reviewer 2: ```suggestion
Zum Beispiel:
```

```js run
function showMessage() {
  alert( 'Hello everyone!' );
}
*!*
showMessage();
*/!*
```

Der Abruf (*call*) `showMessage()` führt den Code der Funktion aus. Darum sehen wir die Nachricht hier zwei Mal.

Dieses Beispeil zeigt perfekt was der hauptsächliche Zweck von Funktionen ist: sie verhindern die Duplikation von Code.

> Falls wir jemals die Nachricht ändern müssten, oder die Art auf die sie angezeigt wird, genügt es den Code einmalig umzuschreiben: der innerhalb der Funktion der ihn ausführt.

Reviewer 3: der innerhalb der Funktion, die ihn ausführt?

> Falls wir jemals die Nachricht ändern müssten oder die Art auf die sie angezeigt wird genügt es den Code einmalig umzuschreiben: der innerhalb der Funktion der ihn ausführt. Genial! 

Reviewer: ```suggestion
Falls wir jemals die Nachricht ändern müssten, oder die Art auf die sie angezeigt wird, genügt es den Code einmalig umzuschreiben: der innerhalb der Funktion der ihn ausführt.
```

> ## Lokale (*local*) Variabeln

Reviewer: ```suggestion
## Lokale (*local*) Variablen
```

> Eine Variabel die innerhalb einer Funktion deklariert wird ist ausschließlich innerhalb dieser sichtbar. 

Reviewer: ```suggestion
Eine Variable, die innerhalb einer Funktion deklariert wird, ist ausschließlich innerhalb dieser sichtbar. 
```

Ein Beispiel:

```js run
function showMessage() {
*!*

>   let message = "Hello, I'm JavaScript!"; // Lokale (*local*) Variabel

Reviewer: ```suggestion
  let message = "Hallo, ich bin JavaScript!"; // Lokale (*local*) Variable
```

*/!*
  alert( message );
}

> showMessage(); // Hello, I'm JavaScript!

Reviewer: ```suggestion
showMessage(); // Hallo, ich bin JavaScript!
```

alert( message ); // <-- Fehler! Die Variabel ist lokal zur Funktion
```

> ## Außenstehende (*outer*) Variabeln

Reviewer: ```suggestion
## Außenstehende (*outer*) Variablen
```

> Eine Funktion kann auch auf außenstehende Variabeln zugreifen wie folgendes Beispiel zeigt:

Reviewer: ```suggestion
Eine Funktion kann auch auf außenstehende Variablen zugreifen, wie folgendes Beispiel zeigt:
```

```js run no-beautify
let *!*userName*/!* = 'John';
function showMessage() {

>   let message = 'Hello, ' + *!*userName*/!*;

Reviewer: ```suggestion
  let message = 'Hallo, ' + *!*userName*/!*;
```

  alert(message);
}

> showMessage(); // Hello, John

Reviewer: ```suggestion
showMessage(); // Hallo, John
```

```

> Die Funktion hat vollständigen Zugriff auf außenstehende Variabeln. Und kann diese auch modifieren.

Reviewer: ```suggestion
Die Funktion hat vollständigen Zugriff auf außenstehende Variablen. Und kann diese auch modifieren.
```

Hier zu sehen:

```js run
let *!*userName*/!* = 'John';

function showMessage() {

>   *!*userName*/!* = "Bob"; // (1) Ändert die außenstehende (*outer*) Variabel 

Reviewer: ```suggestion
  *!*userName*/!* = "Bob"; // (1) Ändert die außenstehende (*outer*) Variable
```

>   let message = 'Hello, ' + *!*userName*/!*;

Reviewer: ```suggestion
  let message = 'Hallo, ' + *!*userName*/!*;
```

  alert(message);
}

alert( userName ); // *!*John*/!* Vor dem Funktionsabruf

showMessage();

alert( userName ); // *!*Bob*/!*, Der Wert wurde von der Funktion modifiziert
```

> Die außenstehende Variabel wird nur genutzt wenn es keine lokale gibt.

Reviewer: ```suggestion
Die außenstehende Variable wird nur genutzt, wenn es keine lokale gibt.
```

> Falls innerhalb der Funktion eine gleichnamige Variabel deklariert wird, wird die außenstehende von dieser überschattet (*shadowed*). Im folgenden Beispiel greift die Funktion auf die lokale Variabel `userName` zu. Die außenstehende wird schlicht ignoriert: 

Reviewer: ```suggestion
Falls innerhalb der Funktion eine gleichnamige Variable deklariert wird, wird die außenstehende von dieser überschattet (*shadowed*). Im folgenden Beispiel greift die Funktion auf die lokale Variable `userName` zu. Die außenstehende wird schlicht ignoriert: 
```

```js run
let userName = 'John';

function showMessage() {
*!*

>   let userName = "Bob"; // deklariert eine lokale Variabel

Reviewer: ```suggestion
  let userName = "Bob"; // deklariert eine lokale Variable
```

*/!*

>   let message = 'Hello, ' + userName; // *!*Bob*/!*

Reviewer: ```suggestion
  let message = 'Hallo, ' + userName; // *!*Bob*/!*
```

  alert(message);
}

> // die Funktion wird ihre eigene Variabel erstellen und nutzen

Reviewer: ```suggestion
// die Funktion wird ihre eigene Variable erstellen und nutzen
```

showMessage();

> alert( userName ); // *!*John*/!*, bleibt unverändert, die Funktion greift nicht auf die außenstehende Variabel zu

Reviewer: ```suggestion
alert( userName ); // *!*John*/!*, bleibt unverändert, die Funktion greift nicht auf die außenstehende Variable zu
```

```

```smart header="*Global* variables"

> Variabeln die außerhalb jeglicher Funktionen deklariert werden, wie das äußere `userName` im oben stehenden Code, nennt man *global*.

Reviewer: ```suggestion
Variablen, die außerhalb jeglicher Funktionen deklariert werden, wie das äußere `userName` im oben stehenden Code, nennt man *global*.
```

> Auf globale Variabeln kann von jeder Funktion aus zugegriffen werden (, solange diese nicht von einer lokalen "überschattet" wird).

Reviewer: ```suggestion
Auf globale Variablen kann von jeder Funktion aus zugegriffen werden (sofern diese nicht von Lokalen "überschattet" werden).
```

> So wenig globale Variabeln wie möglich nutzen ist gute Praxis, denn moderner Code hat wenige bis keine. Die meisten Variabeln verbleiben innerhalb ihrer Funktionen. Manchmal sind globale aber doch praktisch, wenn man bspw. Daten des Projektstands speichern möchte.

Reviewer: ```suggestion
So wenig globale Variablen wie möglich nutzen ist gute Praxis, denn moderner Code hat wenige bis keine. Die meisten Variablen verbleiben innerhalb ihrer Funktionen. Manchmal sind Globale aber doch praktisch, wenn man bspw. Daten des Projektstands speichern möchte.
```

> So wenig globale Variablen wie möglich nutzen ist gute Praxis, denn moderner Code hat wenige bis keine. Die meisten Variablen verbleiben innerhalb ihrer Funktionen. Manchmal sind Globale aber doch praktisch, wenn man bspw. Daten des Projektstands speichern möchte.

Reviewer 2: ```suggestion
Es hat sich als gute und moderne Praxis entwickelt, so wenig globale Variablen wie möglich zu verwenden. Moderner Code hat wenige bis keine dieser globalen Variablen. Die meisten Variablen verbleiben innerhalb ihrer Funktionen. Manchmal sind Globale Variabeln aber doch praktisch, wenn man z.B. Daten des Projektstands speichern möchte.
```

> Es hat sich als gute und moderne Praxis entwickelt, so wenig globale Variablen wie möglich zu verwenden. Moderner Code hat wenige bis keine dieser globalen Variablen. Die meisten Variablen verbleiben innerhalb ihrer Funktionen. Manchmal sind Globale Variabeln aber doch praktisch, wenn man z.B. Daten des Projektstands speichern möchte.

Reviewer 3: "...globale Variablen" instead of  "...Globale Variablen"

```

## Parameter

Wir können Funktionen jedliche Art von Daten zukommen lassen (auch Funktionsargumente [*function arguments*] genannt).

Im unten stehenden Beispiel hat die Funktion zwei Parameter: `from` und `text`.

```js run

> function showMessage(*!*from, text*/!*) { // argumente: from, text

Reviewer: ```suggestion
function showMessage(*!*from, text*/!*) { // Argumente: from, text
```

*/!*
```

> Wenn die Funktion in den Zeilen `(*)` und `(**)`abgerufen wird werden die gegebenen Werte in die lokalen Variabeln `from` und `text` kopiert, woraufhin die Funktion diese nutzt. 

Reviewer: ```suggestion
Wenn die Funktion in den Zeilen `(*)` und `(**)`abgerufen wird, werden die gegebenen Werte in die lokalen Variablen `from` und `text` kopiert, woraufhin die Funktion diese nutzt. 
```

> Hier ist ein weiteres Beispiel: wir haben eine Variabel `from` und spielen diese der Funktion zu. Beachte, dass die Funktion `from` ändert, jedoch diese Änderung außen nicht sichtbar wird weil die Funktion stets eine Kopie des Werts annimmt:

Reviewer: ```suggestion
Hier ist ein weiteres Beispiel: wir haben eine Variable `from` und spielen diese der Funktion zu. Beachte, dass die Funktion `from` ändert, jedoch diese Änderung außen nicht sichtbar wird, weil die Funktion stets eine Kopie des Werts annimmt:
```

*/!*
```

Wenn die Funktion in den Zeilen `(*)` und `(**)`abgerufen wird, werden die gegebenen Werte in die lokalen Variablen `from` und `text` kopiert, woraufhin die Funktion diese nutzt. 

Hier ist ein weiteres Beispiel: wir haben eine Variable `from` und spielen diese der Funktion zu. Beachte, dass die Funktion `from` ändert, jedoch diese Änderung außen nicht sichtbar wird, weil die Funktion stets eine Kopie des Werts annimmt:

```js run
function showMessage(from, text) {

*!*
  from = '*' + from + '*'; // hübscht "from" auf
*/!*

  alert( from + ': ' + text );
}

let from = "Ann";

showMessage(from, "Hallo"); // *Ann*: Hallo

// der Wert von "from" bleibt der selbe, die Funktion hat nur eine lokale Kopie modifiziert
alert( from ); // Ann
```

## Default-Werte

> Wenn ein Parameter nicht gegeben ist wird dessen Wert zu `undefined`.

Reviewer 2: ```suggestion
Wenn ein Parameter nicht gegeben ist, wird dessen Wert zu `undefined`.
```

> Bspw. kann die zuvor genannte Funktion `showMessage(from, text)` m. H. eines einzelen Arguments abgerufen werden:

Reviewer 2: ```suggestion
Bspw. kann die zuvor genannte Funktion `showMessage(from, text)` mit Hilfe eines einzelen Arguments abgerufen werden:
```

```js
showMessage("Ann");
```

Das ist kein Fehler. Ein solcher Abruf würde `"Ann: undefined"` ausgeben, da es kein `text` gibt und daher davon ausgegangen wird das `text === undefined` ist.

Falls wir in diesem Fall ein "default" für `text` festlegen wollen, können wir diesen hinter `=` definieren:

```js run
function showMessage(from, *!*text = "kein Text übergeben"*/!*) {
  alert( from + ": " + text );
}

showMessage("Ann"); // Ann: kein Text übergeben
```
Wenn der `text` Parameter jetzt nicht gegeben wird, erhält er den Wert `"no text given"`

> Hier ist `"no text given"` nur ein String, kann aber auch eine viel komplexere *Expression* sein, die nur evaluiert und zugewissen wird, wenn der entsprechende Parameter fehlt. Das macht folgendes möglich:

Reviewer 2: ```suggestion
Hier ist `"no text given"` nur ein String, kann aber auch eine viel komplexerer Ausdruck (*expression*) sein, der nur evaluiert und zugewissen wird, wenn der entsprechende Parameter fehlt. Das macht folgendes möglich:
```

Author: Entweder man schreibt die englischen Ausdrücke kursiv in den Fließtext
```suggestion
Hier ist `"no text given"` nur ein String, kann aber auch einee viel komplexere *Expression* sein, die nur evaluiert und zugewissen wird, wenn der entsprechende Parameter fehlt. Das macht folgendes möglich:
```

Author: ```suggestion
Hier ist `"no text given"` nur ein String, kann aber auch ein viel komplexerer Ausdruck sein (*expression*) sein, der nur evaluiert und zugewiesen wird, wenn der entsprechende Parameter fehlt. Das macht folgendes möglich:
```

Author: ```suggestion
Hier ist `"no text given"` nur ein String, kann aber auch ein viel komplexerer Ausdruck (*expression*) sein, der nur evaluiert und zugewiesen wird, wenn der entsprechende Parameter fehlt. Das macht folgendes möglich:
```

```js run
function showMessage(from, text = anotherFunction()) {
  // anotherFunction() wird nur ausgeführt wenn kein Text gegeben ist
  // dessen Resultat wird der Wert von *text* sein
}
```

```smart header="Evaluierung von Default-Parametern"
In JavaScript wird ein Default-Parameter immer dann evaluiert, wenn die Funktion ohne den respektiven Parameter abgerufen wird.

Im oberen Beispiel wird `anotherFunction()` jedes Mal abgerufen, wenn `showMessage()` ohne den `text` Parameter abgerufen wird.
```

````smart header="Old-style Default-Parameter"

> In älteren JavaScript-Versionen waren Default-Parameter nicht unterstützt, weshalb sich alternative Wege entwickelt haben um diese trotzdem zu unterstützen. Diese sind meist in alten *scripts* aufzufinden.

Reviewer 2: ```suggestion
In älteren JavaScript-Versionen waren Default-Parameter nicht unterstützt, weshalb sich alternative Wege entwickelt haben. Diese sind meist in alten *scripts* aufzufinden.
```

> showMessage(from, "Hello"); // *Ann*: Hello

Reviewer: ```suggestion
showMessage(from, "Hallo"); // *Ann*: Hallo
```

// der Wert von "from" bleibt der selbe, die Funktion hat nur eine lokale Kopie modifiziert
alert( from ); // Ann
```

## Default-Werte

Wenn ein Parameter nicht gegeben ist wird dessen Wert zu `undefined`.

Bspw. kann die zuvor genannte Funktion `showMessage(from, text)` m. H. eines einzelen Arguments abgerufen werden:

```js
showMessage("Ann");
```

Das ist kein Fehler. Ein solcher Abruf würde `"Ann: undefined"` ausgeben, da es kein `text` gibt und daher davon ausgegangen wird das `text === undefined` ist.

Falls wir in diesem Fall ein "default" für `text` festlegen wollen, können wir diesen hinter `=` definieren:

```js run

> function showMessage(from, *!*text = "no text given"*/!*) {

Reviewer: ```suggestion
function showMessage(from, *!*text = "kein Text übergeben"*/!*) {
```

> showMessage("Ann"); // Ann: no text given

Reviewer: ```suggestion
showMessage("Ann"); // Ann: kein Text übergeben
```

```
Wenn der `text` Parameter jetzt nicht gegeben wird, erhält er den Wert `"no text given"`

> Hier ist `"no text given"` nur ein String, kann aber auch eine viel komplexere *Expression* sein die nur evaluiert und zugewissen wird wenn der entsprechende Parameter fehlt. Das macht folgendes möglich:

Reviewer: ```suggestion
Hier ist `"no text given"` nur ein String, kann aber auch eine viel komplexere *Expression* sein, die nur evaluiert und zugewissen wird, wenn der entsprechende Parameter fehlt. Das macht folgendes möglich:
```

```js run
function showMessage(from, text = anotherFunction()) {

>   // anotherFunction() wird nur ausgeführt wenn kein Text gegebn ist

Reviewer: ```suggestion
  // anotherFunction() wird nur ausgeführt wenn kein Text gegeben ist
```

>   // dessen Resultat wird der Wert von Text sein

Reviewer: ```suggestion
  // dessen Resultat wird der Wert von *text* sein
```

}
```

> ```smart header="Evaluirtung von Default-Parametern"

Reviewer: ```suggestion
```smart header="Evaluierung von Default-Parametern"
```

> In JavaScript wird ein Default-Parameter immer dann evaluiert wenn die Funktion ohne den respektiven Parameter abgerufen wird.

Reviewer: ```suggestion
In JavaScript wird ein Default-Parameter immer dann evaluiert, wenn die Funktion ohne den respektiven Parameter abgerufen wird.
```

Im oberen Beispiel wird `anotherFunction()` jedes Mal abgerufen, wenn `showMessage()` ohne den `text` Parameter abgerufen wird.
```

````smart header="Old-style Default-Parameter"

> In älteren JavaScript-Versionen waren Default-Parameter nicht unterstützt weshalb sich alternative Wege entwickelt haben um diese trotzdem zu unterstützen. Diese sind meist in alten *scripts* aufzufinden.

Reviewer: ```suggestion
In älteren JavaScript-Versionen waren Default-Parameter nicht unterstützt, weshalb sich alternative Wege entwickelt haben um diese trotzdem zu unterstützen. Diese sind meist in alten *scripts* aufzufinden.
```
